[WIP] Fix: Share Price 0 Bug

erin-koen opened the pull request:
After a redemption of all shares, a fund's share price is 0. The metrics service currently will not return an object for months in which every day has a zero share price.

EFK Capital - https://melon.avantgarde.finance/mainnet/fund/0x5d520934de608c30c15946db7cc6270feb1abada
- Share price metrics MTD and QTD n/a, monthly returns showing incorrect and incomplete data (should go [-1.8, 8.3, -, -, -, 24.68, -,- ,-] or similar.


Created 19 Feb 2020
All shares redeemed 14 March 2020
New Investment 17 Aug 2020

`monthend` returns an array of four objects which have the following properties:
| Date | Share Price |
| ----- | ------ |
| 19 Feb | 1 |
| 29 Feb | .98 |
| 31 March | 1.0634 |
| 20 Aug | 6.50 |

the march timestamp is, I assume, the price from the update before I withdrew all assets, because we manipulate dates within the metrics endpoint to show the last second in each month. So it finds the last price in a given month, and then assumes that it's happened on the last day. It doesn't actually do this for the last price of the current month (in this case 20 aug) so I will have to double check that.

My first inclination is that to fix this bug we'll need to change the metrics service to provide month-end prices if the shares are worth 0, but I remember a conversation with @iherger where we chose for it to do this exact thing, so will need to revisit with him to refresh me on the logic behind it and the implications of changing it back. 

Other questions - how do you measure the monthly return in the case where the month end price of the prior month was 0, and now the price is > 0. In the case of EFK capital that'd be July 31 - price zero. August 20 - price 1.3260. I imagine there may be a convention for this but will have to google a bit.

And actually that makes me wonder - how is the share price calculated when investing in a fund where the share price is zero? I.e. I invested 75 MLN and got 6.50 shares when there were no shares outstanding for the fund, so 11.50 MLN/share. Does it go back to 1 ETH/share? Would make sense as I think MLN was 40ish and ETH 450 in USD terms

fubhy commented on src/components/Routes/Fund/FundPerformanceMetrics/FundSharePriceMetrics.tsx:
```diff
@@ -49,7 +49,7 @@ export const FundSharePriceMetrics: React.FC<FundSharePriceMetricsProps> = (prop
     fund.address!
   );
-
+  console.log(monthlyData?.data);
```
Leftover console.log accidentally

fubhy replied:
</trolling>

erin-koen replied:
+eye roll emoji+

erin-koen commented:
I've changed the logic in the `monthlyReturnsFromTimeline` function in `FundMetricsUtilFunctions` to interpret the new timelines from [this PR](https://github.com/avantgardefinance/metrics-service/pull/4) in the metrics service. It sends all timeline items regardless of whether the fund held assets during a given month (previously filtered out timeline items where `calculations.price` was 0). 

The logic i've implemented is commented in the function, but essentially:
- _if the prior month's end price was zero and this month's end price is zero_: this month's return is NA
- _if the prior month's end price was zero and this month's end price is not zero_: set last month's end price in ETH to 1 and run calcs
- _if the prior month's end price was not zero and this months end price is zero_: this month's return is NA
- _if the prior month's end price was not zero and this month's end price is not zero_: do the normal thing

It changes Rhino Fund's first month return by 50ish bps and I'm not entirely certain why. Otherwise it behaves as expected (check EFK Capital).

erin-koen commented:
plz note, there are console.logs and comments throughout

iherger commented on src/components/Routes/Fund/FundPerformanceMetrics/FundMetricsUtilFunctions.ts:
```diff
@@ -47,13 +46,14 @@ export function calculateHoldingPeriodReturns(
   }
 
   const periodStart = relevantPriceData[0];
+  const periodStartPrice = periodStart.calculations.price === 0 ? 1 : periodStart.calculations.price;
```
If the price is zero, then there are no shares, so why would we want to use this as the periodStart?

erin-koen replied:
You're right, this is not correct. take the following case:

-fund created june 15
- first investment july 15

as it's currently written, the qtdReturn will take an eth price of 1, which is fine, but the USDC and WBTC rates it uses to calculate those returns will be from jun30 rather than july 15. 

The issue is that we are not currently returning non-MonthEnd timeline items from this endpoint, they're getting stripped out in the reducer that's calculating monthly returns. I think this will require a call to the range endpoint to get the specific date (first investment). But that gets complicated if a fund gone from zero shares to !zero shares and back again.

My suggestion might be to revert to the original code and open a new PR to deal with this.
